File: Sweep Line/LET_no253_Meeting Rooms II.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        temp = []
        for left, right in intervals:
            temp.append((left, 1))
            temp.append((right, -1))
        
        cur, res = 0, 0
        temp.sort()
        for _, val in temp:
            cur += val
            res = max(res, cur)
        
        return res

class Solution2:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        cur = res = 0
        heap = []
        for start, end in intervals:
            heapq.heappush(heap, (start, 1))
            heapq.heappush(heap, (end, -1))
            
        while heap:
            time, val = heapq.heappop(heap)
            cur += val
            res = max(res, cur)
        
        return res

File: Sweep Line/LET_no253_Meeting Rooms II.cpp
class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        map<int, int> time_uses;
        for (auto& itv : intervals) {
            time_uses[itv[0]]++;
            time_uses[itv[1]]--;
        }

        int cur = 0, res = 0;
        for (auto& it : time_uses) {
            cur += it.second;
            res = max(res, cur);
        }

        return res;
    }
};